### PR TITLE
Don't allow division by false (zero)

### LINKED ISSUE #10717: When indenting code, Godot crashes because of expected indentation
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Godot 2.1.3 stable official, Antergos Linux x86_64, 4.12.8-2-ARCH

**Issue description:**
I added an inbuilt script to a viewport node and copied over the size and stretch override code example from [Godot's docs](https://godot.readthedocs.io/en/stable/learning/features/viewports/viewports.html#doc-viewports) inside the _ready()

It would be copied over unindented as so:
```
extends Viewport

func _ready():
viewport.set_size_override(w,h) #custom size for 2D
viewport.set_size_override_stretch(true/false) #enable stretch for custom size
```

And then, if I indented both lines of code as such:

```
extends Viewport

func _ready():
    viewport.set_size_override(w,h) #custom size for 2D
    viewport.set_size_override_stretch(true/false) #enable stretch for custom size
```

Godot will crash. This does not just apply to the viewport node, but to any node, it seems, as I tried it out on a panel node as well and it led to a crash.

P.S: For some reason, Godot's editor didn't allow code to be copied from it :/

**Steps to reproduce:**
Create a node. Attach a script. Copy/paste the aforementioned code from the Godot docs into the _ready() function and, sometimes without saving, Godot will crash.

THIS DOES NOT HAPPEN IF THE CODE IS WRITTEN BY HAND. (Edit: ok nvm this bit - it's because when writing by hand I didn't include the true/false division argument to the size_override_stretch function...)

### PR DESCRIPTION
**Pull request by hpvb**
This fixes #10717

### CODE REVIEW

**Comment by hpvb** [comment]
@reduz when you have a moment could you have a think about the lhs zero check I added for bool and int?

**Comment by neikeq** [comment]
Would be great to cherry-pick this for 2.1.5

**Comment by hpvb** [comment]
I'd just like @reduz, @akien-mga, or @karroffel  to have a look at the addition of the left-hand-side check. If nobody thinks it's a bad idea I'll happily make a PR for 2.1 also.

**Comment by Faless** [comment]
What is the point of dividing booleans?
```
true / true = ?
true / false = ?
false / true = ?
false / false = ?
```
Shouldn't it just fail when doing that?
On the other side, why checking the value of the left operator?
`0 / 10` is a **valid mathematical expression** which is equal to `0`

**Comment by hpvb** [comment]
@Faless I... didn't know that. I'll remove the lhs check. As for dividing by bools there was already code in int to disallow this so that wasn't added by me.

**Comment by karroffel** [comment]
This should've never been merged. This issue should've never come up. To me it's clear that #10717 had this issue because code was copy pasted that was never meant to be executed.

The "original" author used `true/false` to show that there can be a bool, true **or** false. Unfortunately this is a "correct" expression (which it is not really).

This PR introduced more *weak* typing which is not idiomatic GDScript. Why stop here? Why don't we add `bool / int` or `bool + string`?

In short: this PR does not solve any real problem and is counter intuitive to the way GDScript currently works.

**Comment by hpvb** [comment]
This is not a correct assessment of the situation at all. Check the code right below the code that I edited it already has the guard for division by False. The case 'int / bool' is explicitly handled. I merely added the case 'bool / bool'.

**Comment by Faless** [comment]
The question is, what is the advantage of having int / bool ? (Number/1 or
Number/0)
Why do we need bool / bool (1/1, 1/0, 0/1, 0/0)?
I think arithmetic operations should not be allowed with bool at all and
just throw an error.

On Sep 15, 2017 23:44, "Hein-Pieter van Braam" <notifications@github.com>
wrote:

This is not a correct assessment of the situation at all. Check the code
right below the code that I edited it already has the guard for division by
False. The case 'int / bool' is explicitly handled. I merely added the case
'bool / bool'.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/godotengine/godot/pull/11078#issuecomment-329911355>,
or mute the thread
